File: Machine Learning A-Z New/Part 2 - Regression/Section 6 - Polynomial Regression/polynomial_regression.py
# Polynomial Regression
# lecture 57 https://www.udemy.com/machinelearning/learn/lecture/5813756
# get working directory Spyder
import os
os.getcwd()
# see path of current module (file in use)
import os, sys
os.path.dirname(os.path.abspath(sys.argv[0]))
# set working directory to path of file being used
import os, sys
os.chdir(os.path.dirname(os.path.abspath(sys.argv[0])))

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# =============================================================================
# # Importing the dataset
# =============================================================================
dataset = pd.read_csv('Position_Salaries.csv')
# in our data we only need the years (independent) and salary (dependent variable)
X = dataset.iloc[:, 1].values
# here we have X as a vector, we really want a matrix so we'll use Py against itself
# remembering that the upper value is ignored we add it in
X = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2].values


# =============================================================================
# missing data - none
# =============================================================================

# =============================================================================
# Splitting we don't want to do because the set is so small
# =============================================================================
# Splitting the dataset into the Training set and Test set
"""from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)"""

# =============================================================================
# Encoding - Categorical variables - none
# =============================================================================

# =============================================================================
# Feature Scaling - no here as we didn't do this for the SLR or MLR 
# =============================================================================
# here we are just adding another variable so no need to do this here
"""from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)"""

# =============================================================================
# Fitting Linear Regression to the dataset
# =============================================================================
# we are making LR here to compare to the PLR
from sklearn.linear_model import LinearRegression
# lets make a model, we'll call it lin_regressor
lin_regressor = LinearRegression()
lin_regressor.fit(X, y)

# =============================================================================
# Fitting Polynomial Regression to the dataset
# =============================================================================
# let's get a new class here PolynomialFeatures
from sklearn.preprocessing import PolynomialFeatures
# now we'll make a model with polynomial variables xsquared as in a^2
poly_regressor = PolynomialFeatures(degree = 2)
# in video we started with 2 then it changed to 4
# now we apply the model to the data X, aka we are transforming X into X_poly
X_poly = poly_regressor.fit_transform(X)
# NOW! we'll make the polynomial Y coordinates of our data
#poly_regresor.fit(X_poly, y)
lin_regressor_2 = LinearRegression()
lin_regressor_2.fit(X_poly, y)

# altered the degree to 3 we did this to have things drawn closer to the variables we have
from sklearn.preprocessing import PolynomialFeatures
poly_regressor = PolynomialFeatures(degree = 3)
X_poly = poly_regressor.fit_transform(X)
lin_regressor_2 = LinearRegression()
lin_regressor_2.fit(X_poly, y)

# just for fun let's try 4 - it's predicting well
from sklearn.preprocessing import PolynomialFeatures
poly_regressor = PolynomialFeatures(degree = 4)
X_poly = poly_regressor.fit_transform(X)
lin_regressor_2 = LinearRegression()
lin_regressor_2.fit(X_poly, y)

# =============================================================================
# Visualize
# =============================================================================
# Visualising the Linear Regression results
plt.scatter(X, y, color = 'red')
# we'll add the predictions for Salary
plt.plot(X, lin_regressor.predict(X), color = 'blue')
plt.title('Truth or Bluff (Linear Regression)')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()

# Visualising the Polynomial Regression results
plt.scatter(X, y, color = 'red')
# here we need to use our regression built with the model for our 
# polynomial regression. 
plt.plot(X, lin_regressor_2.predict(poly_regressor.fit_transform(X)), color = 'blue')
plt.title('Truth or Bluff (Polynomial Regression)')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()

# let's do an imrovement to get less straight line action of our curve between points
# Visualising the Polynomial Regression results (for higher resolution and smoother curve)
# what we'll do is adjust X_grid with numpy arange
X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape((len(X_grid), 1))
plt.scatter(X, y, color = 'red')
plt.plot(X_grid, lin_regressor_2.predict(poly_regressor.fit_transform(X_grid)), color = 'blue')
plt.title('Truth or Bluff (Polynomial Regression)')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()

# =============================================================================
# Predicting a new result with Linear Regression
# =============================================================================
# we can pass the regressor model a value to predict a value
lin_regressor.predict([[6.5]])
# Out[84]: array([330378.78787879])  What! ;) this is correct with SLR

# Predicting a new result with Polynomial Regression
lin_regressor_2.predict(poly_regressor.fit_transform([[6.5]]))
# Out[85]: array([158862.4526516])

File: Machine Learning A-Z New/Part 2 - Regression/Section 6 - Polynomial Regression/polynomial_regression.R
# Polynomial Regression in R
getwd() # check Working directory

# Importing the dataset
dataset = read.csv('Position_Salaries.csv')
# in R columns (indexes) start with 1, here we have 1,2,3 columns; Postion, Level #'s, Salary
# we don't care about Position, it's already encoded in the Level column
# so we'll setup Dataset for just 2 and 3
dataset = dataset[2:3]

# We won't do this because we have a very small dataset
# Splitting the dataset into the Training set and Test set
# # install.packages('caTools')
# library(caTools)
# set.seed(123)
# split = sample.split(dataset$Salary, SplitRatio = 2/3)
# training_set = subset(dataset, split == TRUE)
# test_set = subset(dataset, split == FALSE)

# We won't do this because it's not necessary in this dataset
# Feature Scaling
# training_set = scale(training_set)
# test_set = scale(test_set)

# Fitting Linear Regression to the dataset - only for comparing to polynomial
# '.' brings in all elements/variables, and of course there is only one :) 
# we are training on the entire dataset because we didn't split it
lin_regressor = lm(formula = Salary ~ .,
             data = dataset)
summary(lin_regressor)

""" 
RESULTS
Coefficients:
            Estimate Std. Error t value Pr(>|t|)   
(Intercept)  -195333     124790  -1.565  0.15615   
Level          80879      20112   4.021  0.00383 **
"""
# save results to file
write.csv( tidy( lin_regressor ) , "lin_regressor_coefs.csv" )
write.csv( glance( lin_regressor ) , "lin_regressor_an.csv" )

# Fitting Polynomial Regression to the dataset to build the model poly_regressor
# the levels are the degrees. ^2 is squared, ^3 is to the third, etc..
# this adds a new line to the dataset where level is squared
dataset$Level2 = dataset$Level^2
# this add a new line to the dataset where level is to the third X * X * X etc...
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4
# now we rerun our polynomial regression
poly_regressor = lm(formula = Salary ~ .,
              data = dataset)
summary(poly_regressor)

""" 
RESULTS
Coefficients:
             Estimate Std. Error t value Pr(>|t|)   
(Intercept)  184166.7    67768.0   2.718  0.04189 * 
Level       -211002.3    76382.2  -2.762  0.03972 * 
Level2        94765.4    26454.2   3.582  0.01584 * 
Level3       -15463.3     3535.0  -4.374  0.00719 **
Level4          890.2      159.8   5.570  0.00257 ** """
# save results to file
write.csv( tidy( poly_regressor ) , "poly_regressor_coefs.csv" )
write.csv( glance( poly_regressor ) , "poly_regressor_an.csv" )

# Visualising the Linear Regression results
# install.packages('ggplot2')
library(ggplot2)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') +
  geom_line(aes(x = dataset$Level, y = predict(lin_regressor, newdata = dataset)),
            colour = 'blue') +
  ggtitle('Truth or Bluff (Linear Regression)') +
  xlab('Level') +
  ylab('Salary')
# save graph to file
ggsave("lin_graph.pdf")

# Visualising the Polynomial Regression results
# install.packages('ggplot2')
library(ggplot2)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') +
  geom_line(aes(x = dataset$Level, y = predict(poly_regressor, newdata = dataset)),
            colour = 'blue') +
  ggtitle('Truth or Bluff (Polynomial Regression)') +
  xlab('Level') +
  ylab('Salary')
# save graph to file
ggsave("poly_graph.pdf")

# Visualising the Regression Model results (for higher resolution and smoother curve)
# install.packages('ggplot2')
library(ggplot2)
x_grid = seq(min(dataset$Level), max(dataset$Level), 0.1)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') +
  geom_line(aes(x = x_grid, y = predict(poly_regressor,
                                        newdata = data.frame(Level = x_grid,
                                                             Level2 = x_grid^2,
                                                             Level3 = x_grid^3,
                                                             Level4 = x_grid^4))),
            colour = 'blue') +
  ggtitle('Truth or Bluff (Polynomial Regression)') +
  xlab('Level') +
  ylab('Salary')
# save graph to file
ggsave("poly_graph_HD.pdf")

# Predicting a new result with Linear Regression
predict(lin_regressor, data.frame(Level = 6.5))

# Predicting a new result with Polynomial Regression
predict(poly_regressor, data.frame(Level = 6.5,
                             Level2 = 6.5^2,
                             Level3 = 6.5^3,
                             Level4 = 6.5^4))